feat: remove async_trait

washanhanzi opened the pull request:
1. remove async_trait
2. add minimal rust version require to 1.82
3. bump version to 0.6.0

If we can remove `async_trait`, user don't need to use `async_trait` when they impl `RpcFromRequestParts`.

AThilenius commented:
@washanhanzi I tried doing this and remember running into an issue, but now I can't remember what it was. I think because Axum still uses `async_trait` and the types weren't compatible, or something like that. This compiles and works in an actual project for you (as in using it alongside Axum)?

washanhanzi commented:
It's working in my local testing, but I haven't pushed it to production yet. This is my first time using this crate—great work, by the way.

I noticed that axum removed `async_trait` in version [`0.8.0`](https://tokio.rs/blog/2025-01-01-announcing-axum-0-8-0), so I copied the relevant code from there.

https://github.com/tokio-rs/axum/blob/6d78e843982ca299b5d15b7b9095c249e6cdf978/axum-core/src/extract/mod.rs#L56-66

However, I'm not sure why axum requires Rust `v1.78` for the `FromRequestParts` trait. I'll test it with different Rust versions and report back.

AThilenius commented:
I appreciate that! I would love to get rid of `async_trait` so fully support this. It'll definitely be a major revision though, as this will change the public API.

washanhanzi committed:
feat: remove async_trait

1. remove async_trait
2. add minimal rust version require to 1.82
3. bump version to 0.6.0

washanhanzi commented:
I add `1.82` as the minimal required rust version since `icu_properties_data` in `axum-connect-build` require it.
```
package `icu_properties_data v2.0.1` cannot be built because it requires rustc 1.82 or newer

icu_normalizer v2.0.0
└── idna_adapter v1.2.1
    └── idna v1.0.3
        └── url v2.5.4
            └── reqwest v0.12.20
                └── protoc-fetcher v0.1.2
                    └── axum-connect-build v0.5.2 (/Users/jingyu/Github/axum-connect/axum-connect-build)
                        [build-dependencies]
                        └── axum-connect-example v0.3.0 (/Users/jingyu/Github/axum-connect/axum-connect-examples)
```

The `axum-connect` crate has a minimal required rust version of `1.75` because `axum-core`. `axum-connect-build` minimal required rust version is `1.82`. I think it's appropriate to set `axum-connect` minimal version to `1.82`.

With the minimal required rust version, I can remove the conditional compile.


And I think we should commit `Cargo.lock` file.

washanhanzi commented:
> I appreciate that! I would love to get rid of `async_trait` so fully support this. It'll definitely be a major revision though, as this will change the public API.

I forget to mention this, under major version 0, the minor version upgrade is an incompatible update. We only need to bump minor version. If you like, you can give this [LSP](https://github.com/washanhanzi/cargo-appraiser) a try.

AThilenius commented:
@washanhanzi As we already discussed, sorry I dropped the ball on this! I'm going to push this directly to main, as there are changes in the unary/streaming handlers needed as well.
